src/transform/src/normalize_lets.rs

```diff
@@ -0,0 +1,361 @@
+        for (id, expr) in to_emit.iter_mut() {
+            expr.visit_pre_mut(|expr| {
+                if let MirRelationExpr::Get { id, typ } = expr {
+                    if let Id::Local(i) = id {
+                        typ.clone_from(&types[i]);
+                    }
+                }
+            });
+            types.insert(*id, expr.typ());
+        }
+        relation.visit_pre_mut(|expr| {
+            if let MirRelationExpr::Get { id, typ } = expr {
+                if let Id::Local(i) = id {
+                    typ.clone_from(&types[i]);
+                }
+            }
+        });
+
+        // We may need to perform renumbering again if we inlined bindings.
+        // We do this on a full `MirRelationExpr` once we have reintroduced
+        // the `Let` operators from `to_emit`.
+        let renumber = to_emit
+            .keys()
+            .enumerate()
+            .any(|(i, b)| LocalId::new(i as u64) != *b);
+
+        // Reconstitute the stack of let bindings from `to_emit`.
+        for (id, value) in to_emit.into_iter().rev() {
+            *relation = MirRelationExpr::Let {
+                id,
+                value: Box::new(value),
+                body: Box::new(relation.take_dangerous()),
+            };
+        }
+
+        // Do a final pass if we inlined anything, in order to count from zero.
+        if renumber {
+            let mut id_gen = IdGen::default();
+            renumber_bindings(relation, &mut id_gen)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Re-assign an identifier to each `Let`.
+pub fn renumber_bindings(
+    relation: &mut MirRelationExpr,
+    id_gen: &mut IdGen,
+) -> Result<(), crate::TransformError> {
+    let mut renaming = BTreeMap::new();
+    let recursion_guard = RecursionGuard::with_limit(RECURSION_LIMIT);
+    support::renumber_bindings_helper(&recursion_guard, relation, &mut renaming, id_gen)
+}
+
+// Support methods that are unlikely to be useful to other modules.
+mod support {
+
+    use std::collections::BTreeMap;
+
+    use mz_expr::{Id, LocalId, MirRelationExpr};
+    use mz_ore::id_gen::IdGen;
+    use mz_ore::stack::{CheckedRecursion, RecursionGuard};
+
+    // This is pretty simple, but it is used in two locations and seemed clearer named.
+    /// Populates `counts` with the number of uses of each local identifier.
+    pub(super) fn count_local_id_uses(
+        expr: &MirRelationExpr,
+        counts: &mut std::collections::BTreeMap<LocalId, usize>,
+    ) {
+        expr.visit_pre(|expr| {
+            if let MirRelationExpr::Get {
+                id: Id::Local(i), ..
+            } = expr
+            {
+                *counts.entry(*i).or_insert(0) += 1;
+            }
+        });
+    }
+
+    /// Convert an expression containing `Let` bindings into a map from `LocalId` to `Let`-free expressions.
+    ///
+    /// The special key `None` indicates the root of the expression; all other keys are `Some(local_id)`.
```

Comment:
To what key does this comment refer?

Reply:
A historical form of the code no longer evident in the PR. Will fix!